test/client/js/libs/handlebars/ircmessageparser/findLinks.js

```diff
@@ -104,16 +104,150 @@ describe("findLinks", () => {
 		expect(actual).to.deep.equal(expected);
 	});
 
+	it("should find domains without www. but valid tld", () => {
+		const input = "google.com google.lv google.museum";
+		const expected = [{
+			link: "http://google.com",
```

Comment:
Let's be forward and prefix with `https://` instead 🙏 🙏 🙏
I am reading in a few places that today about 75% of the web is HTTPS-ready (and it's only going to go up), so most links posted without a scheme should be fine. And if not, then post the http scheme explicitly.

Reply:
It's not our job to assume https. It is the web servers and browsers (if HSTS is used) to do that. I can imagine this unnecessarily breaking some sites, including local/internal ones.

Reply:
I agree with @astorije on this, http should be the exception. They can specify the scheme explicitly if there's a good reason it *has* to be HTTP. Local/internal is not a good reason - these should be signed by an appropriate development/intranet CA.

Reply:
Everything everywhere has always defaulted to http. HTTPS isn't configured by default.

*All* websites will be available over http (even to just redirect to https) while the opposite is not true. I strongly feel it's not a good reason to break legitimate links that don't have https configured.

No other client (or browser) defaults to https as far as I know, and there's no rfc suggesting that I can find.

Reply:
>Everything everywhere has always defaulted to http

Argumentum ad antiquitatem?

> All websites will be available over http (even to just redirect to https)

This is not the case. All internal services where I work refuse to accept connections on port 80 for instance. It's rare, but completely possible.

> I strongly feel it's not a good reason to break legitimate links that don't have https configured.

This will not impact full, existing legitimate links which are copied. The number of sites it will impact is small (and dropping everyday) and users impacted can easily 

> No other client (or browser) defaults to https as far as I know

Every single browser running HTTPS Everywhere will default to HTTPS where an appropriate ruleset is in place. Non-secure HTTP [is deprecated](https://blog.mozilla.org/security/2015/04/30/deprecating-non-secure-http/) and there are a not insignificant number of features you can't even use without serving your pages securely.

----

I get that backwards-compat. is important, but your argument seems to boil down to "nobody else does it so we shouldn't". In 1981, everyone used plaintext SMTP connections to deliver mail and there was [no RFC suggesting to the contrary](https://tools.ietf.org/html/rfc788#page-16).

Reply:
> It's not our job to assume https. It is the web servers and browsers (if HSTS is used) to do that.

Yet we do assume _a_ scheme, HTTP. To this day and age, assuming HTTPS is sane. Because we do assume a scheme, it's totally our job to choose a sensitive one.

Like @lol768 said, if HTTP is expected, it should be specified. If nothing is specified, it's to our discretion.

> Everything everywhere has always defaulted to http.

I don't agree that this argument justifies going for HTTP. Browsers starting to treat HTTPS as a default and HTTP as insecure goes in this direction.

> _All_ websites will be available over http (even to just redirect to https) while the opposite is not true.

While this is generally the case, nothing technically enforces this. And like @lol768 there are legitimate reasons where this does not apply.

> I strongly feel it's not a good reason to break legitimate links that don't have https configured.

We are not breaking anything by doing this. Links without a scheme are already not valid URLs, so again, it is up to our interpretation. Explicitly specifying `http://` when it is important is the direction the web is heading towards by making things secure by default. My proposal is simply going towards the same direction.

> We're not a major player here, so it would be dumb to do something different.

Major player or not doesn't prevent us to do the _right_ thing. My opinion is still that the right thing to do is to interpret scheme-less links as HTTPS for better security, with extremely limited impact when this (rarely) fails.

> Either way, this entire discussion doesn't belong in this PR. Either make a separate one changing this behaviour or an issue to discuss further.

If I'm not mistaken, this PR introduces the fact that `example.com` is treated as a link, therefore converted to `http://example.com`. My proposal to go for `https://example.com` does belong in this PR, because I do not think we should introduce it as is.

Reply:
> If I'm not mistaken, this PR introduces the fact that `example.com` is treated as a link

I've seen auto-linking bare domains like this go awry, FWIW. For example, clients that happily link every reference to Python or Perl filenames in a programming channel. Is that actually desirable behavior? Especially considering that The Lounge performs link prefetching, and would then try to go and get the page title for a (quite likely) nonexistent domain.

Reply:
I used a known TLD list, so the only question is whether `.py` is filtered against it's second-level domains, as they don't have open registrations (https://en.wikipedia.org/wiki/.py)

If so, that effectively avoids the problem you are describing.

@astorije We also prefix `http` for `www.` links, so this behaviour is not new. And we're still derailing this PR from improving the link matching to one specific small (mis)feature we have. I'd really like it if you created a separate issue to discuss all the details.

Reply:
Good point, @dgw. Like @xPaw said, filtering against known TLDs should help. I'm still expecting some weird 💩 once in a while though, but I'm happy to give it a shot and see if the ratio usefulness/noise ratio is reasonable 😅

> We also prefix `http` for `www.` links, so this behaviour is not new. And we're still derailing this PR

We're not talking about the `www.` previews, so we're not derailing. I'm fine if after this PR we have something not fully consistent as we are introducing a new way to prefetch (but if you want to switch `www.` prefetches to HTTPS as well, be my guest, I would highly support that!)
Current proposal is a one-char change in code (as far as I can tell) + a few chars in tests, not a huge request that derails the PR.

> I'd really like it if you created a separate issue to discuss all the details.

My main concern is that you are against it in the first place. Not doing it in this PR + requesting a new issue effectively looks like a straight no, that will not be reconsidered. I do not want this new feature without HTTPS, and approving this change while hoping of a future issue/PR to do this would be foolish of me.
Heck, it would take me less time to make the change in this PR myself than opening an issue about it (that at this point I have no doubt you would refuse).

🤷‍♂️

Reply:
Would using protocol relative urls (and http for preview fetching as previews don't really care about it) be a good middle ground?